.github/scripts/determine-initial-guest-os-versions.py

```diff
@@ -0,0 +1,148 @@
+import json
+from typing_extensions import Any
+from urllib.request import Request, urlopen
+import urllib.error
+import sys
+
+ROLLOUT_DASHBOARD_ENDPOINT='https://rollout-dashboard.ch1-rel1.dfinity.network/api/v1/rollouts'
+PUBLIC_DASHBOARD_ENDPOINT='https://ic-api.internetcomputer.org/api/v3/subnets?format=json'
+
+# Key definitions
+STATE = 'state'
+FAILED = 'failed'
+COMPLETE = 'complete'
+BATCHES = 'batches'
+SUBNETS = 'subnets'
+GIT_REVISION = 'git_revision'
+EXECUTED_TIMESTAMP_SECONDS = 'executed_timestamp_seconds'
+REPLICA_VERSIONS = 'replica_versions'
+REPLICA_VERSION_ID = 'replica_version_id'
+
+def eprint(*args, **kwargs):
+    print(*args, file=sys.stderr, **kwargs)
+
+def maybe_fetch_data(url : str, timeout : float) -> str | None:
+    """Try fetch data from `url`.
+
+    Will try to fetch data and return `None` if the resource is unavailable for some.
+    If this call fails other resources should be considered.
+    """
+    try:
+        req = Request(url)
+        req.add_header('accept', '*/*')
+        req.add_header('user-agent', 'dfinity-ci')
+        with urlopen(req, timeout=timeout) as response:
+            encoding = response.headers.get_content_charset('utf-8')
+            body = response.read()
+            return body.decode(encoding)
+    except (urllib.error.URLError, urllib.error.HTTPError, TimeoutError) as e:
+        eprint(f"Error fetching data from {url}: {e}")
+        return None
+
+def ensure_json_response(url : str, timeout : float = 10) -> Any | None:
+    response = maybe_fetch_data(url, timeout)
+    if response is None:
+        # Resource was unavailable, if there are
+        # other options they should be considered
+        # at this point
+        return None
+
+    try:
+        return json.loads(response)
+    except (json.JSONDecodeError) as e:
+        # The resource returned some response but it wasn't json.
+        # This error should not be retried and should panic because it was
+        # likely an API change and requires investigation.
+        eprint(f"Received error when decoding the response from {url}: {e}")
+        eprint(f"Response was:\n{response}")
+        exit(1)
+
+def fetch_versions_from_rollout_dashboard() -> list[str] | None:
+    """Fetch data from rollout dashboard
+
+    Panics if the parsed data is not in the expected format.
+    Returns an empty list if the action is retriable.
+    """
+    data = ensure_json_response(ROLLOUT_DASHBOARD_ENDPOINT)
+    if data is None:
+        # Resource was unavailable
+        return []
+
+    versions = set()
+    try:
+        for rollout in data:
+            if rollout.get(STATE, None) is None:
```

Comment:
This is unnecessary guard code.  The structure is not going to change from under you:

https://github.com/dfinity/dre-airflow/tree/main/rollout-dashboard/frontend/src/lib

Reply:
I'd rather it stay like this, to ensure everything is as it should be.